deploy/contracts.go

```diff
@@ -0,0 +1,395 @@
+	// 0: committee witness is not needed
+	// witnessValidators: committee 2/3n+1 with validatorsDeployAllowedContracts
+	// witnessValidatorsAndCommittee: witnessValidators + committee n/2+1 with allowed NNS contract calls
+	deployWitness uint8
+	// contracts that are allowed to be called for the validators-witnessed deployment
+	validatorsDeployAllowedContracts []util.Uint160
+	// additional option for unset tryDeploy to specify deployer of the contract
+	// designated globally. Has no effect if tryDeploy is set.
+	designatedDeployer *util.Uint160
+
+	// optional constructor of extra arguments to be passed into method deploying
+	// the contract. If returns both nil, no data is passed (noExtraDeployArgs can
+	// be used).
+	//
+	// Ignored if tryDeploy is unset.
+	buildExtraDeployArgs func() ([]any, error)
+
+	// address of the Proxy contract deployed in the blockchain. The contract
+	// pays for update transactions.
+	proxyContract util.Uint160
+	// set when syncNeoFSContractPrm relates to Proxy contract. In this case
+	// proxyContract field is unused because address is dynamically resolved within
+	// syncNeoFSContract.
+	isProxy bool
+}
+
+// syncNeoFSContract behaves similar to updateNNSContract but also attempts to
+// deploy the contract if it is missing on the chain and tryDeploy flag is set.
+// If committeeDeployRequired is set, the contract is deployed on behalf of the
+// committee with NNS custom contract scope.
+//
+// Returns address of the on-chain contract synchronized with the record of the
+// NNS domain with parameterized name.
+func syncNeoFSContract(ctx context.Context, prm syncNeoFSContractPrm) (util.Uint160, error) {
+	bLocalNEF, err := prm.localNEF.Bytes()
+	if err != nil {
+		// not really expected
+		return util.Uint160{}, fmt.Errorf("encode local NEF of the contract into binary: %w", err)
+	}
+
+	jLocalManifest, err := json.Marshal(prm.localManifest)
+	if err != nil {
+		// not really expected
+		return util.Uint160{}, fmt.Errorf("encode local manifest of the contract into JSON: %w", err)
+	}
+
+	var proxyCommitteeActor *notary.Actor
+
+	initProxyCommitteeActor := func(proxyContract util.Uint160) error {
+		var err error
+		proxyCommitteeActor, err = newProxyCommitteeNotaryActor(prm.blockchain, prm.localAcc, prm.committee, proxyContract)
+		if err != nil {
+			return fmt.Errorf("create Notary service client sending transactions to be signed by the committee and paid by Proxy contract: %w", err)
+		}
+		return nil
+	}
+
+	if !prm.isProxy {
+		// otherwise, we dynamically receive Proxy contract address below and construct
+		// proxyCommitteeActor after
+		err = initProxyCommitteeActor(prm.proxyContract)
+		if err != nil {
+			return util.Uint160{}, err
+		}
+	}
+
+	// wrap the parent context into the context of the current function so that
+	// transaction wait routines do not leak
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var contractDeployer interface {
+		Sender() util.Uint160
+	}
```

Comment:
Remove it, you need a sender from it in a single place. Just store sender in this variable, that's all.

Reply:
Yes, one place, but different types: `*notary.Actor` and `*actor.Actor`. This _interfaced_ variable is a simple way to cover this case